martinize2/processors/repair_graph.py

```diff
@@ -80,40 +74,131 @@ def make_reference(mol):
+        elif len(matches) > 1:
+            # WARNING
+            print("More than one way to fit {}{} on it's reference. I'm "
+                  "picking one arbitrarily. You might want to fix at least "
+                  "some atomnames.".format(resname, resid))
         match = matches[0]
-        reference_graph.add_node(residx, chain=chain, reference=reference, found=residue, resname=resname, resid=resid, match=match)
+        reference_graph.add_node(residx, chain=chain, reference=reference,
+                                 found=residue, resname=resname, resid=resid,
+                                 match=match)
     reference_graph.add_edges_from(residues.edges())
     return reference_graph
 
 
-def repair_graph(aa_graph, reference_graph):
+def repair_residue(molecule, ref_residue):
+    """
+    Rebuild missing atoms and canonicalize atomnames
     """
-    Repairs a graph ``aa_graph`` produced from a PDB file based on the
-    information in ``reference_graph``. Missing atoms will be reconstructed and
-    atom- and residue names will be canonicalized.
+    # Rebuild missing atoms and canonicalize atomnames
+    missing = []
+    # Step 1: find all missing atoms. Canonicalize names while we're at it.
+    reference = ref_residue['reference']
+    found = ref_residue['found']
+    match = ref_residue['match']
+
+    resid = ref_residue['resid']
+    resname = ref_residue['resname']
+
+    for ref_idx in reference:
+        if ref_idx in match:
+            res_idx = match[ref_idx]
+            node = molecule.nodes[res_idx]
+            # Copy, because it's references everywhere.
+            node['graph'] = molecule.subgraph([res_idx]).copy()
+            node.update(reference.nodes[ref_idx])
+            # Update found as well to keep found and molecule in line. It would
+            # be better to try and figure why found is not a reference, but meh
+            found.nodes[res_idx].update(reference.nodes[ref_idx])
+        else:
+            # if reference.nodes[ref_idx]['element'] != 'H':
+            # INFO
+            print(match)
+            print('Missing atom {}{}:{}'.format(resname, resid, reference.nodes[ref_idx]['atomname']))
+            missing.append(ref_idx)
+    # Step 2: try to add all missing atoms one by one. As long as we added
+    # *something* the situation changed, and we might be able to place another.
+    # We can only place atoms for which we know a neighbour.
+    added = True
+    while missing and added:
+        added = False
+        for ref_idx in missing:
+            # See if the atom we want to add has a known neighbour. Otherwise,
+            # continue to the next.
+            if all(ref_neighbour in missing for ref_neighbour in reference[ref_idx]):
+                continue
+            added = True
+            missing.pop(missing.index(ref_idx))
+            res_idx = max(molecule) + 1  # Alternative: find the first unused number
```

Comment:
Finding the first unused number would be dangerous. It could lead to some outdated references leading somewhere when they should not. The current approach is much safer.

Reply:
Good point.